keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -486,6 +579,8 @@ for our use case.
 The good news is that https://github.com/google/cel-policy-templates-go already has
 demonstrated integrating CEL with OpenAPIv3. We plan to leverage this work.
 
+Support for comparing integers and floats is supported in beta and later (in alpha, ints could only be compared to ints and floats to floats).
```

Comment:
if we're expanding what CEL we accept, that either implies we need a second alpha (so the last alpha and first beta release agree on validation/interpretation of the field) or we need to accept the expanded values over two releases (first release in updates of CRDs that already use the expanded capabilities, second release in all CRD create/update requests)

Reply:
We're changing what CEL will accept in quite a few ways:

1. function library
2. int-or-string handling
3. numeric comparisons
4. CEL expression cost checking
5. Introduction of oldSelf

It would be possible to expand (1), (4) and (5) over two releases.

(2) and (3) are more difficult because they're going to be introduced in CEL, and I wasn't expecting there to be a way to disable/enable them (@TristonianJones do you know?)

For the community, getting CEL to beta is important because we get such limited usage and feedback in alpha. So I'd like to bias our decisions toward things that enable us to get to beta.

Reply:
@jpbetz @liggitt I agree with the assessment that (1), (4), and (5) could be expanded over time, and that (2) and (3) are considered blockers, and deliverables which the CEL team must provide before this feature moves to Beta. Here's a bit more context about what's happening on the CEL side ...

CEL is adding support for heterogeneous numeric comparisons, so you'll be able to expect a comparison of `1 == 1.0` to return `true`, along with ordering operators such as `1 > 0.5`. The other change coming with CEL is `1 == "1"` will return `false` rather than error, making it possible to more easily write expressions like `intOrString == 1 || intOrString == "1"` and have the outcome work as expected.

Since these comparison changes move error results to non-errors, they're considered non-breaking and so we don't plan to explicitly permit them to be enabled / disabled. 

Hopefully that answers your questions?

Reply:
@TristonianJones would it be possible to check if a CEL expression uses heterogeneous types (numeric comparisons or int-or-string)?  For our N-1 compatibility (needed for rollbacks and HA upgrades), we might need to disallow CEL expressions that use the feature for 1 release.

Reply:
Okay, great. Thanks @TristonianJones!  We might not end up needing to check at all for this once since we're leaning toward doing another alpha release and we doing need N-1 compatibility for new features if we go that way. But if we need it, I suspect the disablement option is sufficient.